Add content credentials support

bagasse opened the pull request:
Add content credentials support:
* Add / Modify content credential (GPG Keys and Certs)
* Ability to assign content credential to repository
* Ability to assign content credential to product

mdellweg commented on modules/katello_gpg_key.py:
```diff
@@ -0,0 +1,186 @@
+    server_url: "https://foreman.example.com"
+    name: "RPM-GPG-KEY-my-repo"
+    organization: "Default Organization"
+    content: "-----BEGIN PGP PUBLIC KEY BLOCK-----
```
I preferred, if we did not include an actual gpg-key in this repo.
And for the example, it might be as instructive and much more readable if we did something like:
`content: "{{ lookup('file', 'key.asc') }}"`
I think this applies to the tests, too. Katello does not rigogrously validate, that you insert a valid key, iirc.

mdellweg commented on modules/katello_repository.py:
```diff
@@ -108,8 +112,22 @@
+    url: "http://yum.theforeman.org/plugins/latest/el7/x86_64/"
+    download_policy: background
+
+- name: "Create repository"
```
```suggestion
- name: "Create repository with content credentials"
```

mdellweg commented on modules/katello_product.py:
```diff
@@ -83,6 +87,16 @@
     organization: "My Cool new Organization"
     state: present
+
+- name: "Create CentOS product"
```
```suggestion
- name: "Create CentOS product with content credentials"
```

bagasse commented:
@sean797, I saw that nailgun is using the deprecated gpgkey API ATM, so I used gpg_key for module name but you are right, user facing module should follow naming of current API. Changed module name from `gpg_key` to `content_credential`. Of course it support only gpg keys ATM, but all is ready for supporting other types in the future.

@mdellweg, I've added some tests for products and repos.

mdellweg commented on modules/katello_content_credential.py:
```diff
@@ -0,0 +1,168 @@
+            verify_ssl=dict(type='bool', default=True),
+            name=dict(required=True),
+            organization=dict(required=True),
+            type=dict(default='gpg_key', chhoices=['gpg_key']),
```
```suggestion
            type=dict(default='gpg_key', choices=['gpg_key']),
```

bagasse replied:
Done.

mdellweg commented on test/test_playbooks/tasks/product.yml:
```diff
@@ -12,6 +13,7 @@
     verify_ssl: "{{ foreman_verify_ssl }}"
     name: "{{ product_name }}"
     organization: "{{ organization_name }}"
+    content_credential: "{{ product_content_credential }}"
```
```suggestion
    content_credential: "{{ product_content_credential | omit }}"
```

mdellweg commented on test/test_playbooks/tasks/product.yml:
```diff
@@ -4,6 +4,7 @@
     - organization_name: "Test Organization"
     - product_name: "Test Product"
     - product_description: "A happy little test product"
+    - product_content_credential: ""
```
If you do not put that here, than we can test the behaviour without that option.
`omit` is like it was not specified at all.

mdellweg commented on test/test_playbooks/tasks/repository.yml:
```diff
@@ -5,6 +5,7 @@
     - product_name: "Test Product"
     - repository_name: "Test Repository"
     - repository_content_type: "yum"
+    - repository_content_credential: ""
```
same here

sbernhard commented:
Would be great if you can have a look at https://github.com/SatelliteQE/nailgun/pull/554 if this would help you to support the SSL content credentials, too.

sean797 commented:
Thanks @sbernhard, Nailgun PR is now merged :+1: 

@bagasse mind updating this? Then I think we should be good to merge :+1:

bagasse commented:
@sean797, rebased and updated. Has a new nailgun version is not released yet, I didn't updated all the requirements section in modules documentation for content_credential, product and repository.

sean797 commented on module_utils/ansible_nailgun_cement.py:
```diff
@@ -488,6 +489,11 @@ def find_repository(module, name, product, failsafe=False):
 
 
+def find_content_credential(module, name, organization, failsafe=False):
+    content_credential = ContentCredential(name=name, organization=organization)  # Waiting for Nailgun to use new content credential API
```
I think you can remove this comment?

sean797 commented on modules/katello_content_credential.py:
```diff
@@ -0,0 +1,167 @@
+  - Create and Manage Katello content credentials
+author: "Baptiste Agasse (@bagasse)"
+requirements:
+  - "nailgun >= 0.31.0"
```
Want to change this to 0.32 in anticipation?

bagasse replied:
I've also updated this in product and repository modules as we use content_credential in it too.

sean797 commented on modules/katello_content_credential.py:
```diff
@@ -0,0 +1,167 @@
+    type: bool
+  name:
+    description:
+      - Name of the GPG key
```
Content Credential ;)

sean797 commented on modules/katello_content_credential.py:
```diff
@@ -0,0 +1,167 @@
+            verify_ssl=dict(type='bool', default=True),
+            name=dict(required=True),
+            organization=dict(required=True),
+            content_type=dict(choices=['gpg_key', 'cert']),
```
Docs say `required=True` ?

nailgun defaults to gpg_key, so technically it isn't required but I think I'd prefer if it was

bagasse replied:
Sorry I updated the doc part but forgot this one. For me  nailgun defaults to gpg_key to stay compatible with older versions. Here this is the first version :) . As you I prefer if it is required.

sean797 commented on modules/katello_product.py:
```diff
@@ -155,6 +171,11 @@ def main():
+
+    if 'content_credential' in entity_dict:
+        entity_dict['content_credential'] = find_content_credential(module, name=entity_dict['content_credential'], organization=entity_dict['organization'],
+                                                                    failsafe=True)
```
I think this should be `failsafe=False`? (which is the default)

sean797 commented on modules/katello_repository.py:
```diff
@@ -194,6 +215,10 @@ def main():
 
+    if 'content_credential' in entity_dict:
+        entity_dict['content_credential'] = find_content_credential(module, name=entity_dict['content_credential'], organization=entity_dict['organization'],
+                                                                    failsafe=True)
```
I think this should be `failsafe=False`? (which is the default)

mdellweg commented on README.md:
```diff
@@ -102,6 +102,7 @@ This is a list of modules currently in the repository (please add to the list if
  * foreman_domain: create and maintain domains
  * foreman_job_template: create and maintain job templates and associated template inputs
  * foreman_setting: set and reset settings
+ * katello_content_credential: create and maintain content credential
```
```suggestion
 * katello_content_credential: create and maintain content credentials
```

mdellweg commented on modules/katello_product.py:
```diff
@@ -113,7 +128,7 @@
     'organization': 'organization',
     'description': 'description',
-    # 'gpg_key': 'gpg_key',  # wait for Nailgun
+    'content_credential': 'gpg_key',
```
I have not followed the Nailgun change closely, but is this translation still valid/needed?

mdellweg replied:
On a second thought, I think in the Product/Repository case, you can specify a _gpg key_, a _client cert_, a _server cert_ and a _ca cert_.

So i think we should still call it a `gpg_key` here and add the ssl parts with their names later.
Sorry for bringing that up so late.

mdellweg commented:
@bagasse Thank you!

bagasse commented:
@sean797, @mdellweg, thank you for your time/reviews/patience :+1:
